### DISCUSSION TITLE
Build offline/create Homebrew mirror

### DISCUSSION THREAD

**Discussion by oatbrew** [Homebrew/discussions#2585] [Everyday usage]
When building from source, is it possible to fetch all dependencies and build the binary on an offline system? Or is it possible to create a local mirror of all, or at least the relevant sources?
I think all source files in a formula are referenced  by their name and a checksum. Would it be enough to go through the *.db file, download all the sources manually (on something else than MacOS), transfer them to a MacOS system and put them in some location
where Homebrew will find them and start a build?
Or would there be further dependencies which are not directly visible from the db file and which would be downloaded during a normal build process?

**Comment by carlocab** [comment]
By `*.db` file, do you mean `*.rb`? If so, then yes, there are implicit build dependencies. For example, if the source is an `xz` archive, then `brew` will read that as a build dependency on `xz`.

You can download the sources you need with `brew --fetch --build-from-source`. They will be downloaded into `$(brew --cache)`.

**Reply by oatbrew** [reply]
Yes I meant `*.rb` :) Regarding the implicit build dependencies this would mean the `xz` binary is build first? I'm not sure what the `xz` dependencies look like but wouldn't this result in kind of circular dependencies somehow? Could you catch these kind of nested dependencies? So you run some kind of script across the whole formula tree and you know *all* the sources for building a binary including its dependencies?
But in case this would not be easily possible, copying the files from the cache over to some other system (with the same version of Homebrew) would be enough to get a working build system for a certain binary?

**Reply by carlocab** [reply]
> wouldn't this result in kind of circular dependencies somehow?

I guess if someone made the unfortunate error of using an `xz` archive as the source for the `xz` formula, then yes, this would create a circular dependency. Thankfully we haven't run into this problem yet, and, even if someone tried to update the formula with this change, CI would fail and this mistake would not be merged.

> copying the files from the cache over to some other system (with the same version of Homebrew) would be enough to get a working build system for a certain binary?

The caching mechanism, AFAIK, depends only on the source URL, so, yes, you should just be able to take the contents of `$(brew --cache)` from one machine and dump it into another's `$(brew --cache)` and `brew` should be able to find things just fine.
